pkg/networkservice/common/begin/client.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2021 Cisco and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package begin
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+)
+
+type beginClient struct {
+	Map
+}
+
+// NewClient - provides a new begin client chain element
+func NewClient() networkservice.NetworkServiceClient {
+	return &beginClient{}
+}
+
+func (b *beginClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (conn *networkservice.Connection, err error) {
+	// If we already have a entry from some previous begin.NewServer or begin.NewClient, use that
+	// otherwise, LoadOrStore it and add it to the ctx
+	entry := FromContext(ctx)
+	if entry == nil {
+		entry, _ = b.LoadOrStore(request.GetConnection().GetId(), newEventOriginator(ctx, nil, b))
+		ctx = withEntry(ctx, entry)
+	}
+	recurse := false
```

Comment:
We have a [multiexecutor](https://github.com/networkservicemesh/sdk/blob/main/pkg/tools/multiexecutor/multi_executor.go) for this, probably it would be better to have single `multiexecutor` in a `begin` chain element and create event originators based on it how we do it in serializectx:
https://github.com/networkservicemesh/sdk/blob/77508104772665de3d55ba9fb9a29cb2333f8fa8/pkg/tools/serializectx/context.go#L55-L60